pkg/destroy/azure/azure.go

```diff
@@ -469,7 +469,7 @@ func deleteResourceGroup(ctx context.Context, client resources.GroupsClient, log
 
 	delFuture, err := client.Delete(ctx, name)
 	if err != nil {
-		if wasNotFound(delFuture.Response()) {
+		if delFuture != nil && wasNotFound(delFuture.Response()) {
```

Comment:
I actually found it _doesn't_ make sense -- `delFuture` isn't a pointer. That being the case, I still haven't figured out what could cause the nil pointer exception in the referenced BZ.

Reply:
@2uasimojo you are right. `Delete` returns `GroupsDeleteFuture` which is not a pointer. However it embeds an interface `azure.FutureAPI`.

I guess in the bug you are invesitgating `Delete` returns before `FutureAPI` is instantiated. Probably somewhere here:

https://github.com/openshift/installer/blob/4050517eead6927d299d51b4a57709f56855f878/vendor/github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources/groups.go#L238

or here:

https://github.com/openshift/installer/blob/4050517eead6927d299d51b4a57709f56855f878/vendor/github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources/groups.go#L241-L245

I suspect that it is a client bug and the client itself will be panicing in aa similar fasion if you enable tracing:

https://github.com/openshift/installer/blob/4050517eead6927d299d51b4a57709f56855f878/vendor/github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources/groups.go#L223-L232

Empty value of `GroupsDeleteFuture` has `nil` in the embeded `azure.FutureAPI` hense implicit nil dereference.

Here is the simplified version of what happens:

```go
package main

import (
	"errors"
	"fmt"
	"net/http"
)

func main() {
	simulation()
	fmt.Println("All good")
}

func simulation() {
	delFuture, err := delete()
	if err != nil {
		fmt.Println(delFuture.Response())
		return
	}
	return
}

func delete() (result GroupsDeleteFuture, err error) {
	return result, errors.New("test")
}

type FutureAPI interface {
	Response() *http.Response
}

type GroupsDeleteFuture struct {
	FutureAPI
}

```

Reply:
There should be no check for "Was Not Found" here. If the call to `Delete` returned an error, the error could not have been a "Was Not Found" error. The only way to get a "Was Not Found" error would be after the future completes.

Reply:
See https://github.com/openshift/installer/pull/5314.

Reply:
> in the bug you are invesitgating `Delete` returns before `FutureAPI` is instantiated.

TIL. And it looks like you can actually check the embedded interface for `nil` So a dumb-and-simple fix would be:
```suggestion
		if delFuture.FutureAPI != nil && wasNotFound(delFuture.Response()) {
```
I don't like this very much because it exposes the name of that embedded interface, which is otherwise irrelevant. That's kind of icky for:
- readability: It takes some digging to understand that `.FutureAPI` and `.Response()` are related.
- maintainability: Correct me if I'm wrong, but there's nothing stopping the azure sdk from renaming that interface or replacing it with a different one that implements the same methods. Now we don't compile. (I guess that would wash out in the revendoring process.)

Thoughts?

Reply:
@2uasimojo I have a feeling that #5314 might fix the bug. In the current state of #5314 I see that there is an attempt to walk away from using `delFuture` at all and relying on error returned by the client.

However, it is still worth looking into other places where we have similar pattern. Like here:

https://github.com/openshift/installer/blob/c3a4dbc69860b9b66b019461ce3fc3c546e1914f/pkg/destroy/azure/azure.go#L263-L270

and [here](https://github.com/openshift/installer/blob/c3a4dbc69860b9b66b019461ce3fc3c546e1914f/pkg/destroy/azure/azure.go#L399-L406).

And maybe somewhere else?

Reply:
All other uses of `wasNotFound` in this file are actually being passed responses from clients that aren't returning futures. This is the outlier.